p2p/src/sync/mod.rs

```diff
@@ -89,445 +97,485 @@ where
+                    chainstate::ChainstateEvent::NewTip(block_id, _) => {
+                        let _ = sender.send(block_id).log_err_pfx("The new tip receiver closed");
+                    }
+                },
             );
-        })
+
+        self.chainstate_handle
+            .call_mut(|this| this.subscribe_to_events(subscribe_func))
+            .await
+            .map_err(|_| P2pError::SubsystemFailure)?;
+
+        Ok(receiver)
     }
 
-    /// Unregister peer from the `SyncManager`
-    pub fn unregister_peer(&mut self, peer_id: T::PeerId) {
-        self.peers.remove(&peer_id);
+    async fn handle_message(&mut self, peer: T::PeerId, message: SyncMessage) -> Result<()> {
+        match message {
+            SyncMessage::HeaderListRequest(r) => {
+                self.handle_header_request(peer, r.into_locator()).await
+            }
+            SyncMessage::BlockListRequest(r) => {
+                self.handle_block_request(peer, r.into_block_ids()).await
+            }
+            SyncMessage::HeaderListResponse(r) => {
+                self.handle_header_response(peer, r.into_headers()).await
+            }
+            SyncMessage::BlockResponse(r) => self.handle_block_response(peer, r.into_block()).await,
+        }
     }
 
-    /// Process header request
-    pub async fn process_header_request(
-        &mut self,
-        peer_id: T::PeerId,
-        locator: Locator,
-    ) -> crate::Result<()> {
-        log::debug!("send header response to peer {peer_id}");
-
-        // TODO: check if remote has already asked for these headers?
-        let headers = self.chainstate_handle.call(move |this| this.get_headers(locator)).await??;
-        self.send_header_response(peer_id, headers)
+    /// Processes a header request by sending requested data to the peer.
+    async fn handle_header_request(&mut self, peer: T::PeerId, locator: Locator) -> Result<()> {
+        log::debug!("Headers request from peer {peer}");
+
+        // Check that the peer is connected.
+        self.peers.get(&peer).ok_or(P2pError::PeerError(PeerError::PeerDoesntExist))?;
+
+        if locator.len() > self.p2p_config.msg_max_locator_count.clone().into() {
+            return Err(P2pError::ProtocolError(ProtocolError::LocatorSizeExceeded(
+                locator.len(),
+                self.p2p_config.msg_max_locator_count.clone().into(),
+            )));
+        }
+        log::trace!("locator: {locator:#?}");
+
+        if self.chainstate_handle.call(|c| c.is_initial_block_download()).await?? {
+            // TODO: Check if a peer has permissions to ask for headers during the initial block download.
+            log::debug!("Ignoring headers request because the node is in initial block download");
+            return Ok(());
+        }
+
+        let headers = self.chainstate_handle.call(|c| c.get_headers(locator)).await??;
+        debug_assert!(headers.len() <= self.p2p_config.msg_header_count_limit.clone().into());
```

Comment:
The header-limit in the function c.get_headers() is not taken from p2p. Perhaps you should modify that function to take the limit as a function parameter instead of using the HEADER_LIMIT constant.